src/asset.ts

```diff
@@ -11,28 +11,52 @@ class Asset {
   public ledger: Ledger;
   public name: string;
   public contract?: string;
+  public decimals?: number;
 
-  public constructor(name: string, ledger: Ledger, contract?: string) {
+  public constructor(
+    name: string,
+    ledger: Ledger,
+    contract?: string,
+    decimals?: number
+  ) {
     this.name = name;
     this.ledger = ledger;
     this.contract = contract;
+    this.decimals = decimals;
   }
 }
+export function toAssetFromName(name: string): Asset | undefined {
+  return toAsset({ name });
+}
 
-export function toAsset(asset: string): Asset | undefined {
-  switch (asset) {
+export function toAsset(
+  asset: any,
+  ledger?: Ledger,
+  tokensCreateAsset?: (
+    ledger: Ledger,
+    contractAddress: string
+  ) => Asset | undefined
+): Asset | undefined {
+  if (!asset.name) {
+    return undefined;
+  }
+  switch (asset.name) {
     case Asset.bitcoin.name:
       return Asset.bitcoin;
     case Asset.ether.name:
       return Asset.ether;
     default:
+      if (tokensCreateAsset && ledger && asset.token_contract) {
+        return tokensCreateAsset(ledger, asset.token_contract);
```

Comment:
🔧 If I am not mistaken, given the definition of `Asset`
```
public constructor(
    name: string,
    ledger: Ledger,
    contract?: string,
    decimals?: number
  ) {
    this.name = name;
    this.ledger = ledger;
    this.contract = contract;
    this.decimals = decimals;
  }
```
this line would return as `Asset` with 
```
ledger = ledger
name = asset.token_contract
```
Is that the expected behaviour?

Reply:
`tokensCreateAsset` is defined as follow:

```
(
    ledger: Ledger,
    contractAddress: string
  ) => Asset | undefined
```

It is not related to the constructor of the `Asset` class.

It is actually expected to be `Tokens.createAsset` function. It looks good to me.